### ISSUE TITLE
Custom Application - proxying requests to the external API

### ISSUE THREAD

**Issue by alochowski** [commercetools/merchant-center-application-kit#2634]
I have a question about integrating custom application with my own API - especially genrated JWT token there.
https://docs.commercetools.com/custom-applications/concepts/integrate-with-your-own-api#exchange-token-structure

Is it possible to set **`aud`** param in payload on my own?

Because now this **audience** param is set as provided **`uri`** param in SDK, but our **uri** contain dynamic parts.

We are using https://oauth2-proxy.github.io/oauth2-proxy/ and we expect constant **audience** but for now we have the same as **uri**.

Better for us will be set **audience** as constant value representing e.g. service name or origin URL (protocol, hostname, port) without all path.

Is there any chance to set somehow this **`aud`** in JWT token? Or you are going to add such as possibilities by your SDK?

Regards,
Arek

**Comment by CarlosCortizasCT** [comment]
Hi @alochowski 

Thanks for opening this issue.

I guess we would need a little more background to actually understand what is the problem you are facing.

We know you will send a request from the browser (Custom Application) to the Merchant Center API (configuring it to be forwarded) and we will forward that request to you backend from our gateway sending the JWT token where the `aud` claim will contain the value of the full URL of your endpoint.

It seems that request will first hit you `oauth2-proxy` and then it will get to your backend service where, if you use our [express packages](https://docs.commercetools.com/custom-applications/concepts/integrate-with-your-own-api#validating-the-json-web-token), you could validate the token.

It would be helpful to really understand the steps of your flow, what happens in every step, what causes the problem and what would you need to be different.

Thanks.

**Comment by alochowski** [comment]
Hi @CarlosCortizasCT ,

Thanks for your response.

Everything is as you wrote, except last step - backend service in which we should do validate token.

We don't have implementation on backend side, our backend is security-agnostic.
This is why we have proxy according to the [specification](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3) 
_Each principal intended to process the JWT MUST identify itself with a value in the audience claim._ 

It would be great for us if we will have opportunities to set audience not only per resource (_uri_), but also e.g.  per API or custom string.

**Comment by CarlosCortizasCT** [comment]
Hi @alochowski 

Thanks for your reply.

So, currently we don't support modifying the `aud` claim our gateway uses to create the JWT token that it's used in the proxied request to your backend.

The flow for a proxied request would be something like this example:
![image](https://user-images.githubusercontent.com/97907068/172562584-19f6f63d-da37-4ffe-b36b-80039277c776.png)

For us to potentially think about changes in how we create that claim, I would like to ask what would you need it to look like (please, take the previous example as reference).

I would also like to know where does the JWT validation takes place. Is it in the `oauth2-proxy`? How do you implement the validation? Are you using our [public keys information](https://mc-api.europe-west1.gcp.commercetools.com/.well-known/jwks.json) in the validation step?

**Comment by alochowski** [comment]
Hi @CarlosCortizasCT

Thanks for your reply.

On the `oauth2-proxy` level we are able to set single audience per proxy, so `aud` is expected to be an identifier of the service or API. Therefore, it has to be the same for each request to the same API. So your approach is fine, until there is a path variable in the API call, e.g.: [api.nowhere.com/api/users/ea2a2428-147c-4c3e-90fe-6424cb938fd7](https://api.nowhere.com/api/users/ea2a2428-147c-4c3e-90fe-6424cb938fd7).

To make it working for our current case it would be enough to allow to flag the requests from the browser for Merchant Center Gateway to use `protocol://host:port` part of the requested url in the audience claim, e.g. [api.nowhere.com](https://api.nowhere.com/) (or [api.nowhere.com:8080](https://api.nowhere.com:8080/) if port is used explicitly). 

A fully-functional solution would be to use the requested url until the base path of the API, e.g. [api.nowhere.com/api/users](https://api.nowhere.com/api/users) That would allow to access a number of APIs (backends) behind a common gateway with paths rewrites. However, it requires Merchant Center Gateway to know all the base paths used.

To recap – when requesting [api.nowhere.com/api/users/ea2a2428-147c-4c3e-90fe-6424cb938fd7](https://api.nowhere.com/api/users/ea2a2428-147c-4c3e-90fe-6424cb938fd7) it would be ok if one of the following configurations were be possible:

1.
![Merchant Center Gateway - proxy with static audience1](https://user-images.githubusercontent.com/51888654/172827558-08dbef78-fbbc-42cf-ac4b-2dfbce0c4c34.png)

2.
![Merchant Center Gateway - proxy with static audience2](https://user-images.githubusercontent.com/51888654/172827564-a149c679-d906-4071-a721-5767cb7dde95.png)


The token is validated by oauth2-proxy. As you wrote it is using public keys information. I oauth2-proxy config we could pass the following parameter:

`--extra-jwt-issuers="https://mc-api.europe-west1.gcp.commercetools.com=https://api.nowhere.com/api/users"`


According to [oauth2-proxy documentation](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview/), `extra-jwt-issuers` parameter is “a list of extra JWT issuer=audience […] pairs (where the issuer URL has a .well-known/openid-configuration or a .well-known/jwks.json)”. 
So the public key info is taken exactly from [mc-api.europe-west1.gcp.commercetools.com/.well-known/jwks.json](https://mc-api.europe-west1.gcp.commercetools.com/.well-known/jwks.json), and the proxy expects that the key is signed by the issuer, not expired and the audience matches `https://api.nowhere.com/api/users`.

**Comment by CarlosCortizasCT** [comment]
Hi @alochowski 

Thanks for the detailed explanation.

I guess we now have all the information we need to think about the problem and analyze a potential update in our side to allow such scenarios.

We will let you now when we have any news about it.

**Comment by alochowski** [comment]
Hi @CarlosCortizasCT 

Great :+1:  Thanks a lot.

I will waiting for information.

**Comment by rd-melanie-niermann** [comment]
Hi @CarlosCortizasCT, 

do you have any news here for us, we are blocked right now. 

Thanks and Best 
Melanie

**Comment by CarlosCortizasCT** [comment]
Hi @rd-melanie-niermann, @alochowski 

Regarding the work we planned to do to support use cases like yours, we have good progress regarding what we call "audience policy" (to influence the way the audience claim value is calculated), which you can already start using.

"Audience policy" can have two values:
* `forward-url-full-path`: This is the **default value** and it mimics the current behaviour. The value of the audience will contain the full target URL (origin + pathname)
* `forward-url-origin`: The value of the audience will contain only the origin of the target URL. (**This is the one you would want to use**).

To use the policy you will need to add a new custom header in your requests called `X-Forward-To-Audience-Policy` (with the `forward-url-origin` value).

We continue working on this topic, updating related utilities and documentation (see  #2667 and #2669).
Let us know if this already unblocks you in the current situation.

Thanks.

**Comment by CarlosCortizasCT** [comment]
Hi @alochowski,

We didn't release a new `app-kit` version yet as we are still working on it and we will let you know once the new version is available.

However, you don't need the new version to test the new feature implemented. You just need to add the header with the value mentioned in our previous message.

If you are using our `@commercetools-frontend/sdk` package you would probably some code similar to this:
```
actions.forwardTo.get({
  uri: 'https://api.external-service.com/users/123',
});
```

You would just update it to include the new header:
```
actions.forwardTo.get({
  uri: 'https://api.external-service.com/users/123',
  headers: {
    'X-Forward-To-Audience-Policy': 'forward-url-origin',
  },
});
```

Thanks.

**Comment by alochowski** [comment]
Hi @CarlosCortizasCT,

Yes, I know that.
I checked it, but I still have `aud` in the JWT token with the full path even though this header is set exactly as you wrote.
`actions.forwardTo.get({
uri: 'https://api.external-service.com/users/test',
headers: {
          'X-Forward-To-Audience-Policy': 'forward-url-origin'
        },
      })`
That's why I thought it isn't ready for testing yet until you release a new version.

**Comment by CarlosCortizasCT** [comment]
Hi @alochowski,

Yes, you are right. My provided code snippet does not work, sorry.

We will need to wait for the release of a new version from `app-kit` packages where we got this covered.
We expect to release that new version during this week and we will let you know once its done.

Sorry for the inconvenience.

**Comment by emmenko** [comment]
Hi @alochowski 

We released `v21.8` with the support of the [audience policy](https://docs.commercetools.com/custom-applications/concepts/integrate-with-your-own-api#configuring-the-audience-policy) option.

See also [release notes](https://docs.commercetools.com/custom-applications/releases/2022-07-01-custom-applications-v218) for other useful info about the new release.

**Comment by emmenko** [comment]
Let us know if you still have questions or other problems. Thanks
